LCLSGeneral/LCLSGeneral/POUs/Diagnostics/FB_EtherCATDiag.TcPOU

```diff
@@ -0,0 +1,575 @@
+    fbJson.StartObject();
+    fbJson.AddKey('ecat_master_bAtLeastOneNotInOp');
+    fbJson.AddBool(stMasterDevState.bAtLeastOneNotInOp);
+    fbJson.AddKey('ecat_master_bDcNotInSync');
+    fbJson.AddBool(stMasterDevState.bDcNotInSync);
+    fbJson.AddKey('ecat_master_bDriverNotFound');
+    fbJson.AddBool(stMasterDevState.bDriverNotFound);
+    fbJson.AddKey('ecat_master_bLinkError');
+    fbJson.AddBool(stMasterDevState.bLinkError);
+    fbJson.AddKey('ecat_master_bMissFrmRedMode');
+    fbJson.AddBool(stMasterDevState.bMissFrmRedMode);
+    fbJson.AddKey('ecat_master_bResetActive');
+    fbJson.AddBool(stMasterDevState.bResetActive);
+    fbJson.AddKey('ecat_master_bResetRequired');
+    fbJson.AddBool(stMasterDevState.bResetRequired);
+    fbJson.AddKey('ecat_master_bWatchdogTriggerd');
+    fbJson.AddBool(stMasterDevState.bWatchdogTriggerd);
+    fbJson.AddKey('ecat_master_eEcState');
+    fbJson.AddUdint(stMasterDevState.eEcState);
+    fbJson.EndObject();
+    fbLogger(sMsg:='Master error: error in master device state', eSevr:=TcEventSeverity.Critical, sJson:=fbJson.GetDocument());
+    fbJson.ResetDocument();
+END_IF
+]]></ST>
+    </Implementation>
+    <Action Name="DiagnosticJson" Id="{0fd17872-19c7-46b0-99a6-f83e877544e3}">
+      <Implementation>
+        <ST><![CDATA[fbJson.StartObject();
+fbJson.AddKey('ecat_master_diag');
+
+fbJson.StartArray();
+FOR jsonIdx := 0 to nScannedSlaves DO
+    fbJsonDataType.AddJsonKeyValueFromSymbol(fbJson, 'Slave State Info', 'ST_SlaveStateInfo', 
+    SIZEOF(ST_SlaveStateInfo),
+    ADR(arrSlaveInfo[jsonIdx]));
+END_FOR
+fbJson.EndArray();
+
+
+fbJson.EndObject();
+fbLogger.sJson := fbJson.GetDocument();
+fbJson.ResetDocument();]]></ST>
+      </Implementation>
+    </Action>
+    <Action Name="M_CheckMasterDevState" Id="{c94efdf2-4838-47e0-bb82-f14b06ccb92e}">
+      <Implementation>
+        <ST><![CDATA[(* check master errors based on devstate *)
+bMasterDevStateError 				:= nMasterDevState <> 0;
+stMasterDevState.bLinkError 		:= ((nMasterDevState AND 16#000F) = 1) OR ((nMasterDevState AND 16#000F) = 4);
+stMasterDevState.bResetRequired 	:= ((nMasterDevState AND 16#000F) = 2) OR ((nMasterDevState AND 16#FFF0) = 16#10);
+stMasterDevState.bMissFrmRedMode 	:= (nMasterDevState AND 16#000F) = 8;
+stMasterDevState.bWatchdogTriggerd 	:= (nMasterDevState AND 16#20) = 16#20;
+stMasterDevState.bDriverNotFound 	:= (nMasterDevState AND 16#40) = 16#40;
+stMasterDevState.bResetActive 		:= (nMasterDevState AND 16#80) = 16#80;
+stMasterDevState.bAtLeastOneNotInOp := ((nMasterDevState AND 16#100) = 16#100) OR ((nMasterDevState AND 16#200) = 16#200) OR
+										((nMasterDevState AND 16#400) = 16#400) OR ((nMasterDevState AND 16#800) = 16#800);
+stMasterDevState.bDcNotInSync 		:= (nMasterDevState AND 16#1000) = 16#1000;
+nMasterDevStatePrev 				:= nMasterDevState;
+]]></ST>
+      </Implementation>
+    </Action>
+    <Action Name="M_GetScannedSlaveName" Id="{b8516cb6-a50d-4847-8ef9-760658b0a4c9}">
```

Comment:
I like that these are all broken out into individual and mostly readable actions.